lib/sprockets/npm.rb

```diff
@@ -0,0 +1,57 @@
+# frozen_string_literal: true
+require 'json'
+
+module Sprockets
+  module Npm
+    # Internal: Override resolve_alternates to install package.json behavior.
+    #
+    # load_path    - String environment path
+    # logical_path - String path relative to base
+    #
+    # Returns candiate filenames.
+    def resolve_alternates(load_path, logical_path)
+      candidates, deps = super
+
+      # package.json can only be nested one level deep
+      if !logical_path.index('/'.freeze)
+        dirname = File.join(load_path, logical_path)
+
+        if directory?(dirname)
+          filename = File.join(dirname, 'package.json'.freeze)
+
+          if self.file?(filename)
+            deps << build_file_digest_uri(filename)
+            read_package_main(dirname, filename) do |path|
+              if file?(path)
+                candidates << path
+              end
+            end
+          end
+        end
+      end
+
+      return candidates, deps
+    end
+
+    # Internal: Read package.json's main directive.
+    #
+    # dirname  - String path to component directory.
+    # filename - String path to package.json.
+    #
+    # Returns nothing.
+    def read_package_main(dirname, filename)
+      package = JSON.parse(File.read(filename), create_additions: false)
+
+      case package['main']
+      when String
+        yield File.expand_path(package['main'], dirname)
+      when Array
```

Comment:
Looking at https://nodejs.org/api/modules.html#modules_all_together, I don't think array as the main field is actually acceptable